### PR TITLE
[BEAM-13917] Improve coverage of databaseio package

### PR DESCRIPTION
**Pull request by damccorm**
This adds tests to the dbio read functionality using an in-memory local DB. There are a couple things to consider here:

1) This was the best in-memory go db driver I could find, and it works well for the read tests, but I couldn't get it to test the writes because it throws an error at `SELECT ... WHERE 1 = 0` (called [here](https://github.com/apache/beam/blob/ba2aef5972eca7322ae68f81bafd52d894e43edf/sdks/go/pkg/beam/io/databaseio/database.go#L166) to validate the schema) because the package is still building out full syntactical support. That's why I'm only testing reads, and I don't think we have the option to test writes with this package (since I'd rather not change our implementation to support our mock's bugs).
2) This package isn't incredibly widely used, so there's probably some risk in taking a test dependency on it. If we ever need to upgrade and that's an issue we can always just get rid of the test though. [Package repo](https://github.com/proullon/ramsql) here, its got ~500 stars. It has gotten reasonably consistent patches. Are there requirements from Beam for us to include this package as a dependency?

All in all, I think the pros of getting a full test of the actual functionality (vs testing small pieces of the code) outweigh the cons here, but I'm open to thoughts from others. There's not a great alternative that I see (mocking is basically impossible here since the db is opened from within the DoFn). Its definitely not a straightforward decision to take this change though. 

If we don't do this, I'd probably recommend just skipping databaseio tests.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by damccorm** [comment]
R: @jrmccluskey @riteshghorse @lostluck 

I think this is worth taking on this dependency for the quality of coverage we get, but am interested in thoughts from y'all - it is a less straightforward one.

File: sdks/go/pkg/beam/io/databaseio/database_test.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	if err = insertTestData(db); err != nil {
+		t.Fatalf("Test infra failure: Failed to create/populate table with error %v", err)
+	}
+
+	p, s := beam.NewPipelineWithRoot()
+	elements := Read(s, "ramsql", "user:password@/dbname", "address", reflect.TypeOf(Address{}))
+	passert.Count(s, elements, "NumElements", 2)
+	passert.Equals(s, elements, Address{Street: "orchard lane", Street_number: 1}, Address{Street: "morris st", Street_number: 200})
+
+	if _, err := beam.Run(context.Background(), "direct", p); err != nil {
```
Comment (lostluck) [inline sdks/go/pkg/beam/io/databaseio/database_test.go:50]: Same commentary around using [ptest](https://pkg.go.dev/github.com/apache/beam/sdks/v2@v2.36.0/go/pkg/beam/testing/ptest).

**Review by lostluck — CHANGES_REQUESTED** [review]
LGTM after the ptest change.

**Comment by lostluck** [comment]
To answer the description questions: The risk in taking this dep is minimal:
1. Test dependencies don't get compiled into non-test code.
2. We have it working at version X, and depend on that. Even if they make breaking changes, we'll still be on version X. The main risk is if a user depends on X+N with breaking behaviors and then also runs Our tests, which would hoist things up to X+N.  But it's only in tests, so it still doesn't break user code.

The only other alternative is setting up an in memory SQL lite instance instead. 

Overall, this package should likely be commented and direct users to the Xlang JDBCIO package instead, which will perform better at the moment. We will want to revamp this for Go native implementation eventually, but this is better than nothing.

File: sdks/go/pkg/beam/io/databaseio/database_test.go (hunk 1 above)
Comment (damccorm) [reply sdks/go/pkg/beam/io/databaseio/database_test.go]: Updated!

**Review by lostluck — APPROVED** [review]

**Pull request merged**